File: src/ObjectManager/Object.Tes/FilePacks/Records/345-LIGH.Light.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class LIGHRecord : Record, IHaveEDID, IHaveMODL
    {
        // TESX
        public struct DATAField
        {
            public enum ColorFlags
            {
                Dynamic = 0x0001,
                CanCarry = 0x0002,
                Negative = 0x0004,
                Flicker = 0x0008,
                Fire = 0x0010,
                OffDefault = 0x0020,
                FlickerSlow = 0x0040,
                Pulse = 0x0080,
                PulseSlow = 0x0100
            }

            public float Weight;
            public int Value;
            public int Time;
            public int Radius;
            public ColorRef4 LightColor;
            public int Flags;
            // TES4
            public float FalloffExponent;
            public float FOV;

            public DATAField(UnityBinaryReader r, int dataSize, GameFormatId format)
            {
                if (format == GameFormatId.TES3)
                {
                    Weight = r.ReadLESingle();
                    Value = r.ReadLEInt32();
                    Time = r.ReadLEInt32();
                    Radius = r.ReadLEInt32();
                    LightColor = r.ReadT<ColorRef4>(4);
                    Flags = r.ReadLEInt32();
                    FalloffExponent = 1;
                    FOV = 90;
                    return;
                }
                Time = r.ReadLEInt32();
                Radius = r.ReadLEInt32();
                LightColor = r.ReadT<ColorRef4>(4);
                Flags = r.ReadLEInt32();
                if (dataSize == 32)
                {
                    FalloffExponent = r.ReadLESingle();
                    FOV = r.ReadLESingle();
                }
                else
                {
                    FalloffExponent = 1;
                    FOV = 90;
                }
                Value = r.ReadLEInt32();
                Weight = r.ReadLESingle();
            }
        }

        public override string ToString() => $"LIGH: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL { get; set; } // Model
        public STRVField? FULL; // Item Name (optional)
        public DATAField DATA; // Light Data
        public STRVField? SCPT; // Script Name (optional)??
        public FMIDField<SCPTRecord>? SCRI; // Script FormId (optional)
        public FILEField? ICON; // Male Icon (optional)
        public FLTVField FNAM; // Fade Value
        public FMIDField<SOUNRecord> SNAM; // Sound FormId (optional)

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = r.ReadSTRV(dataSize); return true;
                case "FULL": FULL = r.ReadSTRV(dataSize); return true;
                case "FNAM": if (format != GameFormatId.TES3) FNAM = r.ReadT<FLTVField>(dataSize); else FULL = r.ReadSTRV(dataSize); return true;
                case "DATA":
                case "LHDT": DATA = new DATAField(r, dataSize, format); return true;
                case "SCPT": SCPT = r.ReadSTRV(dataSize); return true;
                case "SCRI": SCRI = new FMIDField<SCPTRecord>(r, dataSize); return true;
                case "ICON":
                case "ITEX": ICON = r.ReadFILE(dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "SNAM": SNAM = new FMIDField<SOUNRecord>(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-LIGH.Light.swift
//
//  LIGHRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class LIGHRecord: Record, IHaveEDID, IHaveMODL {
    // TESX
    public struct DATAField {
        public struct ColorFlags: OptionSet {
            public let rawValue: UInt32
            public static let dynamic = ColorFlags(rawValue: 0x0001)
            public static let canCarry = ColorFlags(rawValue: 0x0002)
            public static let negative = ColorFlags(rawValue: 0x0004)
            public static let flicker = ColorFlags(rawValue: 0x0008)
            public static let fire = ColorFlags(rawValue: 0x0010)
            public static let offDefault = ColorFlags(rawValue: 0x0020)
            public static let flickerSlow = ColorFlags(rawValue: 0x0040)
            public static let pulse = ColorFlags(rawValue: 0x0080)
            public static let pulseSlow = ColorFlags(rawValue: 0x0100)
            
            public init(rawValue: UInt32) {
                self.rawValue = rawValue
            }
        }

        public let weight: Float
        public let value: Int32
        public let time: Int32
        public let radius: Int32
        public let lightColor: ColorRef4
        public let flags: Int32
        // TES4
        public let falloffExponent: Float
        public let fov: Float

        init(_ r: BinaryReader, _ dataSize: Int, _ format: GameFormatId) {
            guard format != .TES3 else {
                weight = r.readLESingle()
                value = r.readLEInt32()
                time = r.readLEInt32()
                radius = r.readLEInt32()
                lightColor = r.readT(4)
                flags = r.readLEInt32()
                falloffExponent = 1
                fov = 90
                return
            }
            time = r.readLEInt32()
            radius = r.readLEInt32()
            lightColor = r.readT(4)
            flags = r.readLEInt32()
            if dataSize == 32 {
                falloffExponent = r.readLESingle()
                fov = r.readLESingle()
            }
            else {
                falloffExponent = 1
                fov = 90
            }
            value = r.readLEInt32()
            weight = r.readLESingle()
        }
    }

    public override var description: String { return "LIGH: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var MODL: MODLGroup? = nil // Model
    public var FULL: STRVField? = nil // Item Name (optional)
    public var DATA: DATAField!  // Light Data
    public var SCPT: STRVField? = nil // Script Name (optional)??
    public var SCRI: FMIDField<SCPTRecord>? = nil // Script FormId (optional)
    public var ICON: FILEField? = nil // Male Icon (optional)
    public var FNAM: FLTVField! // Fade Value
    public var SNAM: FMIDField<SOUNRecord>? = nil // Sound FormId (optional)

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "FULL": FULL = r.readSTRV(dataSize)
        case "FNAM": if format != .TES3 { FNAM = r.readT(dataSize) } else { FULL = r.readSTRV(dataSize) }
        case "DATA",
             "LHDT": DATA = DATAField(r, dataSize, format)
        case "SCPT": SCPT = r.readSTRV(dataSize)
        case "SCRI": SCRI = FMIDField<SCPTRecord>(r, dataSize)
        case "ICON",
             "ITEX": ICON = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "MODT": MODL!.MODTField(r, dataSize)
        case "SNAM": SNAM = FMIDField<SOUNRecord>(r, dataSize)
        default: return false
        }
        return true
    }
}
